x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -133,32 +140,73 @@ public synchronized void closeSession(String sessionUUID) {
         IOUtils.closeWhileHandlingException(restore);
     }
 
+    public synchronized Reader getSessionReader(String sessionUUID, String fileName) throws IOException {
+        RestoreContext restore = onGoingRestores.get(sessionUUID);
+        if (restore == null) {
+            logger.info("could not get session [{}] because session not found", sessionUUID);
+            throw new IllegalArgumentException("session [" + sessionUUID + "] not found");
+        }
+        return restore.getFileReader(fileName);
+    }
+
     private class RestoreContext implements Closeable {
 
         private final String sessionUUID;
         private final IndexShard indexShard;
-        private final Engine.IndexCommitRef commitRef;
+        private final RefCountedCloseable<Engine.IndexCommitRef> session;
+        private RefCountedCloseable<IndexInput> cachedInput;
+        private volatile boolean sessionLocked = false;
+        private Releasable lockRelease = () -> sessionLocked = false;
 
         private RestoreContext(String sessionUUID, IndexShard indexShard, Engine.IndexCommitRef commitRef) {
             this.sessionUUID = sessionUUID;
             this.indexShard = indexShard;
-            this.commitRef = commitRef;
+            this.session = new RefCountedCloseable<>("commit-ref", commitRef);
         }
 
         Store.MetadataSnapshot getMetaData() throws IOException {
             indexShard.store().incRef();
             try {
-                return indexShard.store().getMetadata(commitRef.getIndexCommit());
+                return indexShard.store().getMetadata(session.object.getIndexCommit());
             } finally {
                 indexShard.store().decRef();
             }
         }
 
+        Reader getFileReader(String fileName) throws IOException {
+            if (sessionLocked) {
```

Comment:
maybe just synchronize access on the `Reader.readFileBytes` method to avoid this explicit lock?

Reply:
I have synchronized access on the session object.